Orm/Xtensive.Orm/Reflection/TypeHelper.cs

```diff
@@ -975,16 +971,16 @@ public static bool IsNullable(this Type type) =>
     /// Determines whether given <paramref name="method"/> is a specification
     /// of the provided <paramref name="genericMethodDefinition"/>.
     /// </summary>
-    /// <param name="method">The <see cref="MethodInfo"/> to check.</param>
-    /// <param name="genericMethodDefinition">The <see cref="MethodInfo"/> of the generic method definition
+    /// <param name="methodParams">The <see cref="MethodInfoParams"/> to check.</param>
+    /// <param name="genericMethodDefinitionParams">The <see cref="MethodInfoParams"/> of the generic method definition
     /// to check against.</param>
-    /// <returns><see langword="true"/> if the specified <paramref name="method"/> is a specification
-    /// of the provided <paramref name="genericMethodDefinition"/>.</returns>
-    public static bool IsGenericMethodSpecificationOf(this MethodInfo method, MethodInfo genericMethodDefinition) =>
-      method.MetadataToken == genericMethodDefinition.MetadataToken
-      && (ReferenceEquals(method.Module, genericMethodDefinition.Module)
-        || method.Module == genericMethodDefinition.Module)
-      && method.IsGenericMethod && genericMethodDefinition.IsGenericMethodDefinition;
+    /// <returns><see langword="true"/> if the specified <paramref name="methodParams"/> is a specification
+    /// of the provided <paramref name="genericMethodDefinitionParams"/>.</returns>
+    public static bool IsGenericMethodSpecificationOf(this in MethodInfoParams methodParams, in MethodInfoParams genericMethodDefinitionParams) =>
+      methodParams.MetadataToken == genericMethodDefinitionParams.MetadataToken
+      && methodParams.Module == genericMethodDefinitionParams.Module
+      && methodParams.IsGeneric
+      && genericMethodDefinitionParams.IsGeneric;
```

Comment:
```suggestion
    public static bool IsGenericMethodSpecificationOf(this in MethodInfoParams methodParams, in MethodInfoParams genericMethodDefinitionParams) =>
      genericMethodDefinitionParams.IsGeneric && methodParams == genericMethodDefinitionParams;
```

However maybe `genericMethodDefinitionParams.IsGeneric` is always `true` and not needed. Can we substitute it by some compile-time assertion?

Reply:
Yes, I've moved this checking into constructors of two types: `GenericMethodHandle` & `GenericMethodDefinitionHandle`

Reply:
>  `genericMethodDefinitionParams.IsGeneric && methodParams == genericMethodDefinitionParams`

I'd prefer to avoid using generater record `operator==`, because we want to compare `MetadataToken`s first.
This is `int` very fast comparison and highly likely returns false for most cases.  
The generated operator== is free to compare `Module`s first

Reply:
nit: Now we have two exactly the same structs, which differ only in their `.ctor`. Maybe there is a way to simplify?

Reply:
Looks like structs have optimization in this case if there are no references inside. In fact we have an unmanaged type here :)

Reply:
> Maybe there is a way to simplify?

It is better to have two types for type safety. They are not symmetric

Reply:
> if there are no references inside

But in our case `ModuleHandle` has reference inside